### PR TITLE
Hover effect

### PR DESCRIPTION
**Pull request by gomorizsolt**
Issue #18.

Changes:
- applied @bencefrelli 's flow to the current structure,
- refactored, separated the tooltip related stuff into a single folder,
- extended `GetStyledCalendarElement` with other "generators"(+tests).

Is it possible to test the functions in `Tooltip.js`? I suspect that the correct answer is no because:
- they don't return any value,
- most of the calls are built-in ones.

An additional, slight reason would be that the `showTooltip` and `hideTooltip` functions aren't being exposed.

Unfortunately, the cursor can't be seen on the GIF, but feel free to try the functionality out locally.

pr_23_preview

### CODE REVIEW

File: src/utils/GetStyledCalendarElement/GetStyledCalendarElement.js [hunk 1]
```diff
@@ -76,3 +76,51 @@ export const header = (totalContributions, isLoading) => {
...
+
+  calendarTooltip.id = 'tooltip';
+
+  calendarTooltip.style.background = 'rgba(0, 0, 0, 0.8)';
+  calendarTooltip.style.borderRadius = '3px';
+  calendarTooltip.style.fontSize = '12px';
+  calendarTooltip.style.padding = '10px';
+  calendarTooltip.style.textAlign = 'center';
+  calendarTooltip.style.position = 'absolute';
+  calendarTooltip.style.zIndex = '99999';
```
Comment (thisismydesign) [inline src/utils/GetStyledCalendarElement/GetStyledCalendarElement.js]: How about a just high enough number? :)

File: src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js [hunk 2]
```diff
@@ -109,6 +118,26 @@ describe('TeamContributionCalendar', () => {
         calendarHeader,
       )).to.equal(true);
     });
+
+    it('gets the tooltip element', () => {
```
Comment (thisismydesign) [inline src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: This doesn't test anything useful.

File: src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js [hunk 3]
```diff
@@ -109,6 +118,26 @@ describe('TeamContributionCalendar', () => {
...
+
+    it('appends the tooltip element to the container', () => {
+      teamContributionCalendar.renderActualCalendar();
+
+      expect(calendarContainer.appendChild.calledWithExactly(
+        calendarTooltip,
+      )).to.equal(true);
+    });
+
+    it('adds the tooltip events to the `rect` elements', () => {
```
Comment (thisismydesign) [inline src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: Not what the test does exactly. This sounds like something that could be tested directly.

File: src/utils/GetStyledCalendarElement/GetStyledCalendarElement.js (hunk 1 above)
Comment (gomorizsolt) [reply src/utils/GetStyledCalendarElement/GetStyledCalendarElement.js]: Actually, I copied the same styles that GH uses. :)

But that's right.

File: src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js (hunk 2 above)
Comment (gomorizsolt) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: I don't really get it.

Could you explain your statement above, please?

Comment (thisismydesign) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: Whether `GetStyledCalendarElement.tooltip` was called is not testing the behavior.

Comment (gomorizsolt) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: So it's unnecessary to test function calls whenever there isn't any particular parameter?

Comment (thisismydesign) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: Even with parameters, the use case for testing a function call is that if the call has side effects and those cannot be verified directly. There's no point in testing a getter this way. What should be tested is that the tooltips are there.

Comment (gomorizsolt) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: Cool, thanks for the clarification!

File: src/utils/GetStyledCalendarElement/GetStyledCalendarElement.js (hunk 1 above)
Comment (gomorizsolt) [reply src/utils/GetStyledCalendarElement/GetStyledCalendarElement.js]: It seems that we don't even need the `z-index` property.

File: src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js (hunk 3 above)
Comment (gomorizsolt) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: According to our discussion above, would it be reasonable to remove this case as it doesn't test any behavior related stuff?

Comment (thisismydesign) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: `adds the tooltip events to the rect elements` => regardless of how the test is written now, this is something that tested on the elements, no?

Comment (gomorizsolt) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: Yes, that's it.

But writing `calls addEventsToRectElements` wouldn't fit here either. The solution is to simply omit the case, isn't it?

Comment (thisismydesign) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: can't you test the rendered elements that they have the proper event on them?

Comment (gomorizsolt) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: Not sure, let me take a look into it.

Comment (thisismydesign) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: `aggregateUserCalendars` does have side effects, it updates the calendar.

Comment (gomorizsolt) [reply src/resources/TeamContributionCalendar/TeamContributionCalendar.test.js]: As the testing of `addEventsToRectElements` is pretty complex without the "normal" `document` object, the related test case has been removed.

**Pull request merged**